File: src/code-samples/langchain/nostream-tag.ts
/**
 * Example of using nostream tag to exclude LLM output from the stream.
 */

// :snippet-start: nostream-tag-js
import { ChatAnthropic } from "@langchain/anthropic";
import { StateGraph, StateSchema, START } from "@langchain/langgraph";
import * as z from "zod";

// KEEP MODEL
const streamModel = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });
const internalModel = new ChatAnthropic({
  // KEEP MODEL
  model: "claude-haiku-4-5-20251001",
}).withConfig({
  tags: ["nostream"],
});

const State = new StateSchema({
  topic: z.string(),
  answer: z.string().optional(),
  notes: z.string().optional(),
});

const writeAnswer = async (state: typeof State.State) => {
  const r = await streamModel.invoke([
    { role: "user", content: `Reply briefly about ${state.topic}` },
  ]);
  return { answer: r.content };
};

const internalNotes = async (state: typeof State.State) => {
  // Tokens from this model are omitted from streamMode: "messages" because of nostream
  const r = await internalModel.invoke([
    { role: "user", content: `Private notes on ${state.topic}` },
  ]);
  return { notes: r.content };
};

const graph = new StateGraph(State)
  .addNode("writeAnswer", writeAnswer)
  .addNode("internal_notes", internalNotes)
  .addEdge(START, "writeAnswer")
  .addEdge("writeAnswer", "internal_notes")
  .compile();

const stream = await graph.stream(
  { topic: "AI", answer: "", notes: "" },
  { streamMode: "messages" },
);
// :snippet-end:

// :remove-start:
const streamedNodes: string[] = [];
for await (const [msg, metadata] of stream) {
  if (msg.content) {
    streamedNodes.push(metadata.langgraph_node);
  }
}

if (streamedNodes.includes("internal_notes")) {
  throw new Error(
    "No tokens from the nostream model should appear in the stream",
  );
}

console.log("\n✓ nostream tag example works as expected");
// :remove-end:

File: src/code-samples/langchain/nostream-tag.py
"""Example of using nostream tag to exclude LLM output from the stream."""

# :snippet-start: nostream-tag-py
from typing import Any, TypedDict

from langchain_anthropic import ChatAnthropic
from langgraph.graph import START, StateGraph

# KEEP MODEL
stream_model = ChatAnthropic(model_name="claude-haiku-4-5-20251001")
# KEEP MODEL
internal_model = ChatAnthropic(model_name="claude-haiku-4-5-20251001").with_config(
    {"tags": ["nostream"]}
)


class State(TypedDict):
    topic: str
    answer: str
    notes: str


def answer(state: State) -> dict[str, Any]:
    r = stream_model.invoke(
        [{"role": "user", "content": f"Reply briefly about {state['topic']}"}]
    )
    return {"answer": r.content}


def internal_notes(state: State) -> dict[str, Any]:
    # Tokens from this model are omitted from stream_mode="messages" because of nostream
    r = internal_model.invoke(
        [{"role": "user", "content": f"Private notes on {state['topic']}"}]
    )
    return {"notes": r.content}


graph = (
    StateGraph(State)
    .add_node("write_answer", answer)
    .add_node("internal_notes", internal_notes)
    .add_edge(START, "write_answer")
    .add_edge("write_answer", "internal_notes")
    .compile()
)

initial_state: State = {"topic": "AI", "answer": "", "notes": ""}
stream = graph.stream(initial_state, stream_mode="messages")

# :remove-start:
streamed_nodes: list[str] = []
for msg, metadata in stream:
    if getattr(msg, "content", None) and isinstance(metadata, dict):
        streamed_nodes.append(metadata["langgraph_node"])
assert "internal_notes" not in streamed_nodes, (
    "No tokens from the nostream model should appear in the stream"
)

if __name__ == "__main__":
    print("\n✓ nostream tag example works as expected")
# :remove-end:
# :snippet-end:
